### ISSUE TITLE
[css-text] Circular dependency between letter-spacing and x-height

### ISSUE THREAD

**Issue by litherum** [w3c/csswg-drafts#5498]
https://drafts.csswg.org/css-text-3/#letter-spacing-property

> When the effective spacing between two characters is not zero (due to either justification or a non-zero value of letter-spacing), user agents should not apply optional ligatures

`letter-spacing` takes a `<length>`, which means that it accepts [font-relative units](https://drafts.csswg.org/css-values-4/#font-relative-lengths), like `2ex` for example. Some (all?) browsers can measure the `ex` length by actually rendering a literal "x" character and inspecting its height (at least in the case where the metrics inside the font appear to be bogus). However, the `liga` font feature can change which glyph the "x" character renders with, so if we're supposed to not apply optional ligatures, the height of the "x" character can change depending on whether we're disabling ligatures or not. This means that, if someone applies `letter-spacing` on an element with `ex` units, we may have to measure a glyph, but in order to know which glyph to measure, we have to know whether that glyph has a height of 0.

It's a weird fiddly corner case that probably would never occur in real content, but it's probably worth at least considering how to fix it.

**Comment by litherum** [comment]
Note: Allowing `letter-spacing` to take font-relative units is useful. Also, disabling ligatures when letter-spacing is nonzero is also required for good typography. Whatever solution we pick shouldn't break either of these things.

**Comment by litherum** [comment]
The best solution is probably "don't run shaping when calculating the value of font-relative units."

**Comment by litherum** [comment]
... and the `ch` and `ic` units should be calculated without `letter-spacing` applied (which I think is already the case).

**Comment by faceless2** [comment]
> The best solution is probably "don't run shaping when calculating the value of font-relative units."

Definitely. I would think "liga" is the least of your problems - `salt`, `swsh`, `smcp` and others would have a much bigger impact on units if they were applied, and some of the substitutions could be language dependent too. The metrics really have to be taken from the original glyph as specified by the `glyf` or `CFF` tables.

**Comment by emilio** [comment]
cc @jfkthame

**Comment by fantasai** [comment]
The spec doesn't define exactly how to find the value of an 'ex' from the font, leaves that up to the UA with some pointers. So I don't think there's anything to fix here wrt the specs? https://www.w3.org/TR/css-values-3/#ex

**Comment by frivoal** [comment]
Right, we could arguably be more precise, but given the current lack of precision, no circular dependency seem mandated.

**Comment by fantasai** [comment]
@litherum Shall I close this no change, or is there some change to the spec you want to see here?

**Comment by litherum** [comment]
Can we add a note?
